# Homework Python Programming

- Nama Anggota:
1. Bagas Nur Prayoga
2. Deni Yuniawan
3. Febry Pratama H.S
4. Hilmi Yogantama
5. Lucky Afrida
6. Prambayu Rehyta Muryandari







In [ ]:
import pandas as pd
import numpy as np

## soal 1
Tentukan dari variabel

list_plat_mobil = ['B 1234 AB', 'B 6721 TY', 'B 1233 AY', 'B 6629 DD', 'B 1111 AM', 'B 6726 D', 'D 11223 KJ', 'AE 44677 GH', 'AE 67269 AA']

Ada berapa jumlah plat genap dan ganjil.

In [ ]:
list_plat_mobil = ['B 1234 AB', 'B 6721 TY', 'B 1233 AY', 'B 6629 DD', 'B 1111 AM', 'B 6726 D', 'D 11223 KJ', 'AE 44677 GH', 'AE 67269 AA']

list_ganjil = []
list_genap = []

for plat in list_plat_mobil:
    angka_plat = ''
    for huruf in plat:
        if huruf.isdigit():
            angka_plat += huruf

    angka_plat = int(angka_plat)
    if angka_plat % 2 == 0:
        list_genap.append(plat)
    else:
        list_ganjil.append(plat)

print("Jumlah plat genap :", len(list_genap))
print("Jumlah plat ganjil:", len(list_ganjil))

Jumlah plat genap : 2
Jumlah plat ganjil: 7


## soal 2

Buatlah deret fibonacci 12 angka yang dimulai dari 2 dan 3 menggunakan iterasi cara for atau while.

Output yang diharapkan adalah [2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377]

In [ ]:
deret = [2, 3]
while len(deret) < 13:
    next_deret = deret[-1] + deret[-2]
    deret.append(next_deret)

print("Deret fibonacci:", deret)

Deret fibonacci: [2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610]


## soal 3

Team Campaign PT ABC akan membuat promo bundling category product (berisi 3 category setiap bundling) untuk promo 17 Agustus.  PT ABC memiliki 10 category product.  Ketentuan Bundling tidak boleh hanya memiliki 1 jenis category dalam 1 bundling dan urutan tidak menjadi faktor pembeda (c1 - c2 - c3 = c2 - c1 - c3). Tentukan ada berapa jumlah kemungkinan bundling yang bisa terbentuk?

Berikut list code category <b>list_id_category</b> = [c1, c2, c3, c4, c5, c6, c7 , c8, c9, c10]

In [ ]:
##tuliskode
list_id_category = ['c'+str(i) for i in range(1,11)]
combination_list = []

for cprod1 in list_id_category:
  for cprod2 in list_id_category:
    for cprod3 in list_id_category:
      combination = {cprod1, cprod2, cprod3}
      if len(combination) == 3 and combination not in combination_list:
        combination_list.append(combination)



print(f'Jumlah kombinasi produk adalah: {len(combination_list)}')
display(combination_list)

[{'c1', 'c2', 'c3'},
 {'c1', 'c2', 'c4'},
 {'c1', 'c2', 'c5'},
 {'c1', 'c2', 'c6'},
 {'c1', 'c2', 'c7'},
 {'c1', 'c2', 'c8'},
 {'c1', 'c2', 'c9'},
 {'c1', 'c10', 'c2'},
 {'c1', 'c3', 'c4'},
 {'c1', 'c3', 'c5'},
 {'c1', 'c3', 'c6'},
 {'c1', 'c3', 'c7'},
 {'c1', 'c3', 'c8'},
 {'c1', 'c3', 'c9'},
 {'c1', 'c10', 'c3'},
 {'c1', 'c4', 'c5'},
 {'c1', 'c4', 'c6'},
 {'c1', 'c4', 'c7'},
 {'c1', 'c4', 'c8'},
 {'c1', 'c4', 'c9'},
 {'c1', 'c10', 'c4'},
 {'c1', 'c5', 'c6'},
 {'c1', 'c5', 'c7'},
 {'c1', 'c5', 'c8'},
 {'c1', 'c5', 'c9'},
 {'c1', 'c10', 'c5'},
 {'c1', 'c6', 'c7'},
 {'c1', 'c6', 'c8'},
 {'c1', 'c6', 'c9'},
 {'c1', 'c10', 'c6'},
 {'c1', 'c7', 'c8'},
 {'c1', 'c7', 'c9'},
 {'c1', 'c10', 'c7'},
 {'c1', 'c8', 'c9'},
 {'c1', 'c10', 'c8'},
 {'c1', 'c10', 'c9'},
 {'c2', 'c3', 'c4'},
 {'c2', 'c3', 'c5'},
 {'c2', 'c3', 'c6'},
 {'c2', 'c3', 'c7'},
 {'c2', 'c3', 'c8'},
 {'c2', 'c3', 'c9'},
 {'c10', 'c2', 'c3'},
 {'c2', 'c4', 'c5'},
 {'c2', 'c4', 'c6'},
 {'c2', 'c4', 'c7'},
 {'c2', 'c4', 'c8'},
 {'c

Jumlah kombinasi produk adalah: 120


## soal 4

Buatlah sebuah function string_processor dengan inputan 2 parameter, yaitu suatu string 1 kata dan nama proses yang akan dilakukan. Function ini memiliki 4 proses
- first_char : return huruf pertama dari string yang diinput
- last_char : return huruf terakhir dari string yang diinput
- num_chars : return jumlah huruf string input
- num_vowels : return jumlah huruf vocal pada string input


In [ ]:
##tuliskode
def string_processor(input_string : 'str', process = 'first_char'):
  #membuang spasi pada awal dan akhir string
  input_string = input_string.strip()

  #process pada string input
  list_vowel = ['a', 'i', 'u', 'e', 'o']
  if process == 'first_char':
    result = input_string[0]
  elif process == 'last_char':
    result = input_string[-1]
  elif process == 'num_chars':
    list_huruf = [char for char in input_string if char != " "]
    result = len(list_huruf)
  elif process == 'num_vowels':
    list_vowel_in_input_string = [char for char in input_string if (char != " ") & (char in list_vowel)]
    result = len(list_vowel_in_input_string)

  #menampilkan hasil
  return result

In [ ]:
string_processor('Kelompok I Data Science Bootcamp Rakamin Batch 45')

'K'

In [ ]:
string_processor('Kelompok I Data Science Bootcamp Rakamin Batch 45', process = 'last_char')

'5'

In [ ]:
string_processor('Kelompok I Data Science Bootcamp Rakamin Batch 45', process = 'num_vowels')

15

In [ ]:
string_processor('Kelompok I Data Science Bootcamp Rakamin Batch 45', process = 'num_chars')

42

## soal 5

Menggunakan dataset rakamin_store_city_monthly! Buatlah 1 kolom baru pada dengan teknik apply lambda tanpa function def (cukup one liner) dengan nama kolom is_gmail dengan value True jika menggunakan gmail dan False jika selain gmail.

Tentukan ada berapa user yang menggunakan email gmail pada tahun 2018!

In [ ]:
#run baris kode ini lalu upload file rakamin_store_city_monthly.csv dari lokal drive ke google collab ini
from google.colab import files
uploaded = files.upload()

Saving rakamin_store_city_monthly.csv to rakamin_store_city_monthly.csv


In [ ]:
from datetime import datetime
df_rakamin_store = pd.read_csv('rakamin_store_city_monthly.csv', sep=';')
df_rakamin_store['is_gmail']= df_rakamin_store['email'].apply(lambda mail: True if '@gmail' in mail else False)
df_rakamin_store['month'] = pd.to_datetime(df_rakamin_store['month'], format='%Y-%m')
df_rakamin_store['year'] = df_rakamin_store['month'].dt.year
gmail_2018 = df_rakamin_store[(df_rakamin_store['year'] == 2018) & (df_rakamin_store['is_gmail'] == True)]
print( "jumlah user pengguna gmail pada tahun 2018 ada ",len(gmail_2018), "orang")

jumlah user pengguna gmail pada tahun 2018 ada  3 orang


## soal 6

Menggunakan dataset superstore! Buatlah function yang bisa menghitung Jumlah Sales dan Profit pada suatu tahun dan region tertentu dengan default tahun 2017. Hint : Buat kolom tahun

<b>Tentukan total sales dan profit:</b><br>
- South Region pada tahun 2016
- West Region pada tahun 2017

In [ ]:
#run baris kode ini lalu upload file superstore.xls disini dari lokal drive ke google collab ini
from google.colab import files
uploaded = files.upload()

Saving superstore.xls to superstore.xls


In [ ]:
superstore = pd.read_excel('superstore.xls')

def hitung_sales_profit(data, region, tahun=2017):
    # Copy data
    data_used = data.copy()

    # Filter berdasarkan region
    condition_1 = (data_used['Region'] == region)
    data_used = data_used[condition_1]

    # Buat tambahan kolom tahun
    data_used['Order Year'] = data_used['Order Date'].dt.year

    # Filter berdasarkan tahun
    condition_2 = (data_used['Order Year'] == tahun)
    data_used = data_used[condition_2]

    # Hitung agregasi
    result = data_used.groupby(['Region', 'Order Year'], as_index=False).agg(
        total_sales=('Sales', 'sum'),
        total_profit=('Profit', 'sum'),
    )

    # Memberikan nilai kembalian saat memanggil fungsi
    return result

# Panggil fungsi hitung_sales_profit dengan kondisi yang diberikan
hasil_perhitungan = hitung_sales_profit(superstore, region='South', tahun=2016)
hasil_perhitungan2 = hitung_sales_profit(superstore, region='West', tahun=2017)

# Gabungkan hasil perhitungan dari kedua region
hasil_gabungan = pd.concat([hasil_perhitungan, hasil_perhitungan2], ignore_index=True)

print(hasil_gabungan)

  Region  Order Year  total_sales  total_profit
0  South        2016   93610.2235    17702.8084
1   West        2017  250128.3655    43808.9561


## soal 7

Menggunakan dataset superstore! Buatlah 3 kelompok 'segment'  (high, medium, dan low) berdasarkan performance pembelian yang digunakan sebagai acuan tim bisnis untuk mengelompokan customer superstore. <br>

<b>Instruksi:</b>
1. Bantulah tim bisnis untuk menentukan kriteria pengelompokan untuk 3 segmentasi customer ini, dan gunakan iterrows atau apply lambda sebagai teknik pembuatan kolom baru!
2. Berikan alasan kalian dalam penentuan kriteria pengelompokan customer yang dibuat!
3. Tentukan jumlah customer pada masing-masing segment di tahun 2017!

In [ ]:
#run baris kode ini lalu upload file superstore.xls disini dari lokal drive ke google collab ini
from google.colab import files
uploaded = files.upload()

In [ ]:
df_superstore = pd.read_excel('superstore.xls')
df_superstore.head(3)

Row ID        Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11  Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16  Second Class    DV-13045   

     Customer Name    Segment        Country         City  ... Postal Code  \
0      Claire Gute   Consumer  United States    Henderson  ...       42420   
1      Claire Gute   Consumer  United States    Henderson  ...       42420   
2  Darrin Van Huff  Corporate  United States  Los Angeles  ...       90036   

   Region       Product ID         Category Sub-Category  \
0   South  FUR-BO-10001798        Furniture    Bookcases   
1   South  FUR-CH-10000454        Furniture       Chairs   
2    West  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.96         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.62         2   

   Discount    Profit  
0       0.0   41.9136  
1       0.0  219.5820  
2       0.0    6.8714  

[3 rows x 21 columns]

step1: ekstraksi tahun

In [ ]:
#get year
df_superstore['year'] = df_superstore['Order Date'].dt.year #menggunakan fungsi .dt.year untuk mengambil data tahun dari kolom order_date
df_superstore['year'].tail (3)

9991    2017
9992    2017
9993    2017
Name: year, dtype: int32

step2: cek dataset

In [ ]:
#get year
min_date = df_superstore['Order Date'].min() #min date
max_date = df_superstore['Order Date'].max() #max date
#min_date
max_date

Timestamp('2017-12-30 00:00:00')

In [ ]:
#persebaran order per Year
df_superstore['year'].value_counts()

year
2017    3312
2016    2587
2015    2102
2014    1993
Name: count, dtype: int64

In [ ]:
#persebaran order per Region
df_superstore['Region'].value_counts()

Region
West       3203
East       2848
Central    2323
South      1620
Name: count, dtype: int64

step3: Agregasi

*   Agregasi dilakukan pada kolom yang paling penting yaitu sales, customer_id, customer_name dan tahun.
*   Agregasi dilakukan pada kolom2 tersebut agar mendapatkan nilai total penjualan(sales) per pelanggan untuk tiap tahun.


In [ ]:
df_superstore_group_year = df_superstore.groupby(['year','Customer ID', 'Customer Name'])['Sales'].sum().reset_index()
df_superstore_group_year.tail()

year Customer ID       Customer Name     Sales
2494  2017    XP-21865        Xylona Preis   449.312
2495  2017    YC-21895      Yoseph Carroll   750.680
2496  2017    YS-21880       Yana Sorensen  5340.264
2497  2017    ZC-21910    Zuschuss Carroll   227.066
2498  2017    ZD-21925  Zuschuss Donatelli    61.440

step 4: Statistik Deskriptif
 berdasarkan total penjualan pertahun yaitu dari data kolom year dan kolom sales

In [ ]:
df_superstore_group_year_desc = df_superstore_group_year.groupby(['year'])['Sales'].describe().reset_index()
df_superstore_group_year_desc

year  count         mean          std    min      25%      50%        75%  \
0  2014  595.0   813.861341  1558.339428  1.869   86.448  321.552   951.7495   
1  2015  573.0   821.173663  1115.847785  1.720  120.564  389.434  1065.9200   
2  2016  638.0   954.867708  1524.135920  1.964  107.271  495.729  1210.8170   
3  2017  693.0  1058.030671  1463.910713  1.188  204.405  604.910  1334.6416   

         max  
0  23661.228  
1   9022.324  
2  18344.052  
3  14203.278

step 5:


*  Data asli hasil agregasi (hasil dari step3) dimerge/gabung dengan data statistik deskriptif total penjualan(hasil dari step 4), digabungkan berdasarkan kolom year dan disimpan dalam variable **df_superstore_group_year**
*  Dimana data statistik deskriptif  yang ditampilkan hanya 'year', 'mean', percentile percentile '25%', percentile '50%', percentile '75%'
*   Penggabungan ini dilakukan dengan metode inner, yang berarti hanya data yang memiliki tahun yang sama di kedua DataFrame yang akan digabung.




In [ ]:
df_superstore_group_year = df_superstore_group_year.merge(df_superstore_group_year_desc[['year', 'mean', '25%', '50%', '75%']], on='year', how='inner')
df_superstore_group_year.head()

year Customer ID  Customer Name     Sales        mean     25%      50%  \
0  2014    AA-10315     Alex Avila   756.048  813.861341  86.448  321.552   
1  2014    AA-10375   Allen Armold    50.792  813.861341  86.448  321.552   
2  2014    AA-10480   Andrew Allen    27.460  813.861341  86.448  321.552   
3  2014    AA-10645  Anna Andreadi  1434.330  813.861341  86.448  321.552   
4  2014    AB-10015  Aaron Bergman   322.216  813.861341  86.448  321.552   

        75%  
0  951.7495  
1  951.7495  
2  951.7495  
3  951.7495  
4  951.7495

Step 6: segmentasi customer

*   Segmentasi dibuat berdasarkan olah data statistik deskriptif dan hasil agregasi pada total penjualan pertahun yaitu data kolom sales, year
*   karena data total penjualan/sales adalah ukuran kinerja utama bidang usaha pertokoan (store)
*   segmentasi dibagi jadi 3 low, mid, high:
1.   low --> total sales per tahun pada percentile **25%**
2.   medium --> total sales per tahun pada percentile **50%**
3.   medium --> total sales per tahun pada percentile **75%**




In [ ]:
#function segmentation user besed on total sales yearly
def segment(x):
    if x['Sales'] < x['50%']:
        segment = 'Low'
    elif x['Sales'] <= x['75%'] and x['Sales'] >= x['50%']:
        segment = 'Mid'
    else:
        segment = 'High'
    return segment

Step 7: function Lambda untuk menambahkan kolom baru yaitu : segment, dimana kolom segment adalah informasi segmentasi total penjualan

In [ ]:
df_superstore_group_year['segment'] = df_superstore_group_year.apply(lambda x: segment(x), axis=1)
df_superstore_group_year.head()

year Customer ID  Customer Name     Sales        mean     25%      50%  \
0  2014    AA-10315     Alex Avila   756.048  813.861341  86.448  321.552   
1  2014    AA-10375   Allen Armold    50.792  813.861341  86.448  321.552   
2  2014    AA-10480   Andrew Allen    27.460  813.861341  86.448  321.552   
3  2014    AA-10645  Anna Andreadi  1434.330  813.861341  86.448  321.552   
4  2014    AB-10015  Aaron Bergman   322.216  813.861341  86.448  321.552   

        75% segment  
0  951.7495     Mid  
1  951.7495     Low  
2  951.7495     Low  
3  951.7495    High  
4  951.7495     Mid

Step 8: total jumlah customer pada masing-masing segment di tahun 2017

In [ ]:
#filter tahun 2017 dan count unique based on customer id
df_superstore_group_year[(df_superstore_group_year.year == 2017)].groupby(['segment'])['Customer ID'].nunique()

segment
High    173
Low     346
Mid     174
Name: Customer ID, dtype: int64

📖rangkuman:
1.  penentuan kriteria pengelompokan untuk 3 segmentasi
customer berdasarkan data **total sales pertahun** fungsi lambda digunakan untuk membuat kolom baru yang berisi data segmentasi
2.  alasan dalam total sales kriteria pengelompokan customer adalah karena data total penjualan/sales adalah ukuran kinerja utama bidang usaha pertokoan (store)
3.  total jumlah customer pada masing-masing segment di tahun 2017 sebagai berikut:
*   High    173
*   Mid     174
*   Low     346

